// ring-client-api/lib/streaming/peer-connection.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WeriftPeerConnection = void 0;
/* eslint-disable brace-style */
const werift_1 = require("werift");
const rxjs_1 = require("rxjs");
const util_1 = require("../util");
const subscribed_1 = require("../subscribed");
const ringIceServers = [
    'stun:stun.kinesisvideo.us-east-1.amazonaws.com:443',
    'stun:stun.kinesisvideo.us-east-2.amazonaws.com:443',
    'stun:stun.kinesisvideo.us-west-2.amazonaws.com:443',
    'stun:stun.l.google.com:19302',
    'stun:stun1.l.google.com:19302',
    'stun:stun2.l.google.com:19302',
    'stun:stun3.l.google.com:19302',
    'stun:stun4.l.google.com:19302',
];
class WeriftPeerConnection extends subscribed_1.Subscribed {
    constructor() {
        super();
        this.onAudioRtp = new rxjs_1.Subject();
        this.onAudioRtcp = new rxjs_1.Subject();
        this.onVideoRtp = new rxjs_1.Subject();
        this.onVideoRtcp = new rxjs_1.Subject();
        this.onIceCandidate = new rxjs_1.Subject();
        this.onConnectionState = new rxjs_1.ReplaySubject(1);
        this.returnAudioTrack = new werift_1.MediaStreamTrack({ kind: 'audio' });
        this.onRequestKeyFrame = new rxjs_1.Subject();
        const pc = (this.pc = new werift_1.RTCPeerConnection({
            codecs: {
                audio: [
                    new werift_1.RTCRtpCodecParameters({
                        mimeType: 'audio/opus',
                        clockRate: 48000,
                        channels: 2,
                    }),
                    new werift_1.RTCRtpCodecParameters({
                        mimeType: 'audio/PCMU',
                        clockRate: 8000,
                        channels: 1,
                        payloadType: 0,
                    }),
                ],
                video: [
                    new werift_1.RTCRtpCodecParameters({
                        mimeType: 'video/H264',
                        clockRate: 90000,
                        rtcpFeedback: [
                            { type: 'transport-cc' },
                            { type: 'ccm', parameter: 'fir' },
                            { type: 'nack' },
                            { type: 'nack', parameter: 'pli' },
                            { type: 'goog-remb' },
                        ],
                        parameters: 'packetization-mode=1;profile-level-id=640029;level-asymmetry-allowed=1',
                    }),
                    new werift_1.RTCRtpCodecParameters({
                        mimeType: 'video/rtx',
                        clockRate: 90000,
                    }),
                ],
            },
            iceServers: ringIceServers.map((server) => ({ urls: server })),
            iceTransportPolicy: 'all',
            bundlePolicy: 'disable',
        })), audioTransceiver = pc.addTransceiver(this.returnAudioTrack, {
            direction: 'sendrecv',
        }), videoTransceiver = pc.addTransceiver('video', {
            direction: 'recvonly',
        });
        audioTransceiver.onTrack.subscribe((track) => {
            track.onReceiveRtp.subscribe((rtp) => {
                this.onAudioRtp.next(rtp);
            });
            track.onReceiveRtcp.subscribe((rtcp) => {
                this.onAudioRtcp.next(rtcp);
            });
            track.onReceiveRtp.once(() => {
                (0, util_1.logDebug)('received first audio packet');
            });
        });
        videoTransceiver.onTrack.subscribe((track) => {
            track.onReceiveRtp.subscribe((rtp) => {
                this.onVideoRtp.next(rtp);
            });
            track.onReceiveRtcp.subscribe((rtcp) => {
                this.onVideoRtcp.next(rtcp);
            });
            track.onReceiveRtp.once(() => {
                (0, util_1.logDebug)('received first video packet');
                this.addSubscriptions((0, rxjs_1.merge)(this.onRequestKeyFrame, (0, rxjs_1.interval)(4000)).subscribe(() => {
                    videoTransceiver.receiver
                        .sendRtcpPLI(track.ssrc)
                        .catch((e) => (0, util_1.logError)(e));
                }));
                this.requestKeyFrame();
            });
        });
        this.pc.onIceCandidate.subscribe((iceCandidate) => {
            this.onIceCandidate.next(iceCandidate);
        });
        pc.iceConnectionStateChange.subscribe(() => {
            (0, util_1.logInfo)(`iceConnectionStateChange: ${pc.iceConnectionState}`);
            if (pc.iceConnectionState === 'closed') {
                this.onConnectionState.next('closed');
            }
        });
        pc.connectionStateChange.subscribe(() => {
            (0, util_1.logInfo)(`connectionStateChange: ${pc.connectionState}`);
            this.onConnectionState.next(pc.connectionState);
        });
    }
    async createOffer() {
        const offer = await this.pc.createOffer();
        await this.pc.setLocalDescription(offer);
        return offer;
    }
    async acceptAnswer(answer) {
        await this.pc.setRemoteDescription(answer);
    }
    addIceCandidate(candidate) {
        return this.pc.addIceCandidate(candidate);
    }
    requestKeyFrame() {
        this.onRequestKeyFrame.next();
    }
    close() {
        this.pc.close().catch(util_1.logError);
        this.unsubscribe();
    }
}
exports.WeriftPeerConnection = WeriftPeerConnection;
