// ring-client-api/lib/streaming/simple-webrtc-session.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SimpleWebRtcSession = void 0;
const util_1 = require("../util");
function liveViewUrl(path) {
    return `https://api.ring.com/integrations/v1/liveview/${path}`;
}
class SimpleWebRtcSession {
    constructor(camera, restClient) {
        this.camera = camera;
        this.restClient = restClient;
        this.sessionId = (0, util_1.generateUuid)();
    }
    async start(sdp) {
        const response = await this.restClient.request({
            method: 'POST',
            url: liveViewUrl('start'),
            json: {
                session_id: this.sessionId,
                device_id: this.camera.id,
                sdp,
                protocol: 'webrtc',
            },
        });
        return response.sdp;
    }
    async end() {
        const resp = await this.restClient.request({
            method: 'POST',
            url: liveViewUrl('end'),
            json: {
                session_id: this.sessionId,
            },
        });
        return resp;
    }
    async activateCameraSpeaker() {
        await this.restClient.request({
            method: 'PATCH',
            url: liveViewUrl('options'),
            json: {
                session_id: this.sessionId,
                actions: ['turn_off_stealth_mode'],
            },
        });
    }
}
exports.SimpleWebRtcSession = SimpleWebRtcSession;
