// ring-client-api/lib/streaming/webrtc-connection.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebrtcConnection = void 0;
const ws_1 = require("ws");
const rxjs_1 = require("rxjs");
const operators_1 = require("rxjs/operators");
const util_1 = require("../util");
const peer_connection_1 = require("./peer-connection");
const subscribed_1 = require("../subscribed");
class WebrtcConnection extends subscribed_1.Subscribed {
    constructor(ticket, camera, options) {
        super();
        this.camera = camera;
        this.onSessionId = new rxjs_1.ReplaySubject(1);
        this.onOfferSent = new rxjs_1.ReplaySubject(1);
        this.dialogId = (0, util_1.generateUuid)();
        this.onCameraConnected = new rxjs_1.ReplaySubject(1);
        this.onCallAnswered = new rxjs_1.ReplaySubject(1);
        this.onCallEnded = new rxjs_1.ReplaySubject(1);
        this.onError = new rxjs_1.ReplaySubject(1);
        this.onMessage = new rxjs_1.ReplaySubject();
        this.sessionId = null;
        this.hasEnded = false;
        this.ws = new ws_1.WebSocket(`wss://api.prod.signalling.ring.devices.a2z.com:443/ws?api_version=4.0&auth_type=ring_solutions&client_id=ring_site-${(0, util_1.generateUuid)()}&token=${ticket}`, {
            headers: {
                // This must exist or the socket will close immediately but content does not seem to matter
                'User-Agent': 'android:com.ringapp',
            },
        });
        if (options.createPeerConnection) {
            // we were passed a custom peer connection factory
            this.pc = options.createPeerConnection();
            // passing rtp packets is not supported for custom peer connections
            this.onAudioRtp = new rxjs_1.Subject();
            this.onVideoRtp = new rxjs_1.Subject();
        }
        else {
            // no custom peer connection factory, use the werift and pass along rtp packets
            const pc = new peer_connection_1.WeriftPeerConnection();
            this.pc = pc;
            this.onAudioRtp = pc.onAudioRtp;
            this.onVideoRtp = pc.onVideoRtp;
        }
        this.onWsOpen = (0, rxjs_1.fromEvent)(this.ws, 'open');
        const onMessage = (0, rxjs_1.fromEvent)(this.ws, 'message'), onError = (0, rxjs_1.fromEvent)(this.ws, 'error'), onClose = (0, rxjs_1.fromEvent)(this.ws, 'close');
        this.addSubscriptions(onMessage
            .pipe((0, operators_1.concatMap)((event) => {
            const message = JSON.parse(event.data);
            this.onMessage.next(message);
            return this.handleMessage(message).catch((e) => {
                if (e instanceof Error &&
                    e.message.includes('negotiate codecs failed')) {
                    e = new Error('Failed to negotiate codecs.  This is a known issue with Ring cameras.  Please see https://github.com/dgreif/ring/wiki/Streaming-Legacy-Mode');
                }
                this.onError.next(e);
                throw e;
            });
        }))
            .subscribe(), onError.subscribe((e) => {
            (0, util_1.logError)(e);
            this.callEnded();
        }), onClose.subscribe(() => {
            this.callEnded();
        }), this.pc.onConnectionState.subscribe((state) => {
            if (state === 'failed') {
                (0, util_1.logError)('Stream connection failed');
                this.callEnded();
            }
            if (state === 'closed') {
                (0, util_1.logDebug)('Stream connection closed');
                this.callEnded();
            }
        }), this.onError.subscribe((e) => {
            (0, util_1.logError)(e);
            this.callEnded();
        }), this.onWsOpen.subscribe(() => {
            const connectionType = camera.isRingEdgeEnabled ? 'Ring Edge' : 'Cloud';
            (0, util_1.logDebug)(`WebSocket connected for ${camera.name} (${connectionType})`);
            this.initiateCall().catch((e) => {
                (0, util_1.logError)(e);
                this.callEnded();
            });
        }), 
        // The ring-edge session needs a ping every 5 seconds to keep the connection alive
        (0, rxjs_1.interval)(5000).subscribe(() => {
            this.sendSessionMessage('ping');
        }), this.pc.onIceCandidate.subscribe(async (iceCandidate) => {
            await (0, rxjs_1.firstValueFrom)(this.onOfferSent);
            this.sendMessage({
                method: 'ice',
                dialog_id: this.dialogId,
                body: {
                    doorbot_id: camera.id,
                    ice: iceCandidate.candidate,
                    mlineindex: iceCandidate.sdpMLineIndex,
                },
            });
        }));
    }
    async initiateCall() {
        const { sdp } = await this.pc.createOffer();
        this.sendMessage({
            method: 'live_view',
            dialog_id: this.dialogId,
            body: {
                doorbot_id: this.camera.id,
                stream_options: { audio_enabled: true, video_enabled: true },
                sdp,
            },
        });
        this.onOfferSent.next();
    }
    async handleMessage(message) {
        if (message.body.doorbot_id !== this.camera.id) {
            // ignore messages for other cameras
            return;
        }
        if (['session_created', 'session_started'].includes(message.method) &&
            'session_id' in message.body &&
            !this.sessionId) {
            this.sessionId = message.body.session_id;
            this.onSessionId.next(this.sessionId);
        }
        if (message.body.session_id && message.body.session_id !== this.sessionId) {
            // ignore messages for other sessions
            return;
        }
        switch (message.method) {
            case 'session_created':
            case 'session_started':
                // session already stored above
                return;
            case 'sdp':
                await this.pc.acceptAnswer(message.body);
                this.onCallAnswered.next(message.body.sdp);
                this.activate();
                return;
            case 'ice':
                await this.pc.addIceCandidate({
                    candidate: message.body.ice,
                    sdpMLineIndex: message.body.mlineindex,
                });
                return;
            case 'pong':
                return;
            case 'notification':
                const { text } = message.body;
                if (text === 'camera_connected') {
                    this.onCameraConnected.next();
                    return;
                }
                else if (text === 'PeerConnectionState::kConnecting' ||
                    text === 'PeerConnectionState::kConnected') {
                    return;
                }
                break;
            case 'close':
                (0, util_1.logError)('Video stream closed');
                (0, util_1.logError)(message.body);
                this.callEnded();
                return;
        }
        (0, util_1.logError)('UNKNOWN MESSAGE');
        (0, util_1.logError)(message);
    }
    sendSessionMessage(method, body = {}) {
        const sendSessionMessage = (sessionId) => {
            const message = {
                method,
                dialog_id: this.dialogId,
                body: {
                    ...body,
                    doorbot_id: this.camera.id,
                    session_id: sessionId,
                },
            };
            this.sendMessage(message);
        };
        if (this.sessionId) {
            // Send immediately if we already have a session id
            // This is needed to send `close` before closing the websocket
            sendSessionMessage(this.sessionId);
        }
        else {
            // Otherwise wait for the session id to be set and then send the message
            this.addSubscriptions(this.onSessionId.pipe((0, operators_1.take)(1)).subscribe(sendSessionMessage));
        }
    }
    sendMessage(message) {
        if (this.hasEnded) {
            return;
        }
        this.ws.send(JSON.stringify(message));
    }
    sendAudioPacket(rtp) {
        if (this.hasEnded) {
            return;
        }
        if (this.pc instanceof peer_connection_1.WeriftPeerConnection) {
            this.pc.returnAudioTrack.writeRtp(rtp);
        }
        else {
            throw new Error('Cannot send audio packets to a custom peer connection implementation');
        }
    }
    activate() {
        (0, util_1.logInfo)('Activating Session');
        // the activate_session message is required to keep the stream alive longer than 70 seconds
        this.sendSessionMessage('activate_session');
        this.sendSessionMessage('stream_options', {
            audio_enabled: true,
            video_enabled: true,
        });
    }
    activateCameraSpeaker() {
        // Fire and forget this call so that callers don't get hung up waiting for answer (which might not happen)
        this.addSubscriptions(this.onCameraConnected.pipe((0, operators_1.take)(1)).subscribe(() => {
            this.sendSessionMessage('camera_options', {
                stealth_mode: false,
            });
        }));
    }
    callEnded() {
        if (this.hasEnded) {
            return;
        }
        try {
            this.sendMessage({
                reason: { code: 0, text: '' },
                method: 'close',
            });
            this.ws.close();
        }
        catch (_) {
            // ignore any errors since we are stopping the call
        }
        this.hasEnded = true;
        this.unsubscribe();
        this.onCallEnded.next();
        this.pc.close();
    }
    stop() {
        this.callEnded();
    }
    requestKeyFrame() {
        this.pc.requestKeyFrame?.();
    }
}
exports.WebrtcConnection = WebrtcConnection;
