// werift/lib/rtp/src/extra/processor/rtcpCallback.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RtcpSourceCallback = void 0;
const rx_mini_1 = __importDefault(require("rx.mini"));
class RtcpSourceCallback {
    constructor() {
        Object.defineProperty(this, "cb", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "destructor", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "onStopped", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: new rx_mini_1.default()
        });
        Object.defineProperty(this, "input", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: (rtcp) => {
                if (this.cb) {
                    this.cb({ rtcp });
                }
            }
        });
        Object.defineProperty(this, "destroy", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: () => {
                if (this.destructor) {
                    this.destructor();
                    this.destructor = undefined;
                }
                this.cb = undefined;
                this.onStopped.allUnsubscribe();
            }
        });
    }
    toJSON() {
        return {};
    }
    pipe(cb, destructor) {
        this.cb = cb;
        this.destructor = destructor;
        return this;
    }
    stop() {
        if (this.cb) {
            this.cb({ eol: true });
        }
        this.onStopped.execute();
    }
}
exports.RtcpSourceCallback = RtcpSourceCallback;
