// ring-client-api/lib/ring-chime.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RingChime = void 0;
const ring_types_1 = require("./ring-types");
const rest_client_1 = require("./rest-client");
const rxjs_1 = require("rxjs");
const settingsWhichRequireReboot = [
    'ding_audio_id',
    'ding_audio_user_id',
    'motion_audio_id',
    'motion_audio_user_id',
];
class RingChime {
    constructor(initialData, restClient) {
        this.initialData = initialData;
        this.restClient = restClient;
        this.onRequestUpdate = new rxjs_1.Subject();
        this.id = this.initialData.id;
        this.deviceType = this.initialData.kind;
        this.model = ring_types_1.ChimeModel[this.deviceType] || 'Chime';
        this.onData = new rxjs_1.BehaviorSubject(this.initialData);
    }
    updateData(update) {
        this.onData.next(update);
    }
    requestUpdate() {
        this.onRequestUpdate.next(null);
    }
    get data() {
        return this.onData.getValue();
    }
    get name() {
        return this.data.description;
    }
    get description() {
        return this.data.description;
    }
    get volume() {
        return this.data.settings.volume;
    }
    getRingtones() {
        return this.restClient.request({
            url: (0, rest_client_1.clientApi)('ringtones'),
        });
    }
    async getRingtoneByDescription(description, kind) {
        const ringtones = await this.getRingtones(), requestedRingtone = ringtones.audios.find((audio) => audio.available &&
            audio.description === description &&
            audio.kind === kind);
        if (!requestedRingtone) {
            throw new Error('Requested ringtone not found');
        }
        return requestedRingtone;
    }
    chimeUrl(path = '') {
        return (0, rest_client_1.clientApi)(`chimes/${this.id}/${path}`);
    }
    playSound(kind) {
        return this.restClient.request({
            url: this.chimeUrl('play_sound'),
            method: 'POST',
            json: { kind },
        });
    }
    async snooze(time) {
        // time is in minutes, max 24 * 60 (1440)
        await this.restClient.request({
            url: this.chimeUrl('do_not_disturb'),
            method: 'POST',
            json: { time },
        });
        this.requestUpdate();
    }
    async clearSnooze() {
        await this.restClient.request({
            url: this.chimeUrl('do_not_disturb'),
            method: 'POST',
        });
        this.requestUpdate();
    }
    async updateChime(update) {
        await this.restClient.request({
            url: this.chimeUrl(),
            method: 'PUT',
            json: { chime: update },
        });
        this.requestUpdate();
        // inform caller if this change requires a reboot
        return Object.keys(update.settings || {}).some((key) => settingsWhichRequireReboot.includes(key));
    }
    setVolume(volume) {
        if (volume < 0 || volume > 11) {
            throw new Error(`Volume for ${this.name} must be between 0 and 11, got ${volume}`);
        }
        return this.updateChime({
            settings: {
                volume,
            },
        });
    }
    async getHealth() {
        const response = await this.restClient.request({
            url: (0, rest_client_1.clientApi)(`chimes/${this.id}/health`),
        });
        return response.device_health;
    }
}
exports.RingChime = RingChime;
