// ring-client-api/lib/util.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.clearTimeouts = clearTimeouts;
exports.delay = delay;
exports.logDebug = logDebug;
exports.logInfo = logInfo;
exports.logError = logError;
exports.useLogger = useLogger;
exports.enableDebug = enableDebug;
exports.generateUuid = generateUuid;
exports.getHardwareId = getHardwareId;
exports.requestInput = requestInput;
exports.stringify = stringify;
exports.mapAsync = mapAsync;
exports.randomInteger = randomInteger;
exports.randomString = randomString;
exports.fromBase64 = fromBase64;
exports.toBase64 = toBase64;
exports.buildSearchString = buildSearchString;
const debug_1 = __importDefault(require("debug"));
const colors_1 = require("colors");
const readline_1 = require("readline");
const uuid_1 = require("uuid");
const systeminformation_1 = require("systeminformation");
const logger_1 = __importDefault(require("@eneris/push-receiver/dist/utils/logger"));
const debugLogger = (0, debug_1.default)('ring'), uuidNamespace = 'e53ffdc0-e91d-4ce1-bec2-df939d94739d';
let logger = {
    logInfo(message) {
        debugLogger(message);
    },
    logError(message) {
        debugLogger((0, colors_1.red)(message));
    },
}, debugEnabled = false;
const timeouts = new Set();
function clearTimeouts() {
    timeouts.forEach((timeout) => {
        clearTimeout(timeout);
    });
}
function delay(milliseconds) {
    return new Promise((resolve) => {
        const timeout = setTimeout(() => {
            timeouts.delete(timeout);
            resolve(undefined);
        }, milliseconds);
        timeouts.add(timeout);
    });
}
function logDebug(message) {
    if (debugEnabled) {
        logger.logInfo(message);
    }
}
function logInfo(...message) {
    logger.logInfo(...message);
}
function logError(message) {
    logger.logError(message);
}
function useLogger(newLogger) {
    logger = newLogger;
}
function enableDebug() {
    debugEnabled = true;
}
function generateUuid(seed) {
    if (seed) {
        return (0, uuid_1.v5)(seed, uuidNamespace);
    }
    return (0, uuid_1.v4)();
}
async function getHardwareId(systemId) {
    if (systemId) {
        return generateUuid(systemId);
    }
    const timeoutValue = '-1', { os: id } = await Promise.race([
        (0, systeminformation_1.uuid)(),
        delay(5000).then(() => ({ os: timeoutValue })),
    ]);
    if (id === timeoutValue) {
        logError('Request for system uuid timed out.  Falling back to random session id');
        return (0, uuid_1.v4)();
    }
    if (id === '-') {
        // default value set by systeminformation if it can't find a real value
        logError('Unable to get system uuid.  Falling back to random session id');
        return (0, uuid_1.v4)();
    }
    return generateUuid(id);
}
async function requestInput(question) {
    const lineReader = (0, readline_1.createInterface)({
        input: process.stdin,
        output: process.stdout,
    }), answer = await new Promise((resolve) => {
        lineReader.question(question, resolve);
    });
    lineReader.close();
    return answer.trim();
}
function stringify(data) {
    if (typeof data === 'string') {
        return data;
    }
    if (typeof data === 'object' && Buffer.isBuffer(data)) {
        return data.toString();
    }
    return JSON.stringify(data) + '';
}
function mapAsync(records, asyncMapper) {
    return Promise.all(records.map((record) => asyncMapper(record)));
}
function randomInteger() {
    return Math.floor(Math.random() * 99999999) + 100000;
}
function randomString(length) {
    const uuid = generateUuid();
    return uuid.replace(/-/g, '').substring(0, length).toLowerCase();
}
// Override push receiver logging to avoid ECONNRESET errors leaking
function logPushReceiver(...args) {
    try {
        if (args[0].toString().includes('ECONNRESET')) {
            // don't log ECONNRESET errors
            return;
        }
    }
    catch (_) {
        // proceed to log error
    }
    logDebug('[Push Receiver]');
    logDebug(args[0]);
}
logger_1.default.error = logPushReceiver;
function fromBase64(encodedInput) {
    const buff = Buffer.from(encodedInput, 'base64');
    return buff.toString('ascii');
}
function toBase64(input) {
    const buff = Buffer.from(input);
    return buff.toString('base64');
}
function buildSearchString(search) {
    return ('?' +
        Object.entries(search)
            .filter(([, value]) => value !== undefined && value !== null)
            .map(([key, value]) => `${key}=${value}`)
            .join('&'));
}
